### ISSUE TITLE
Add auto backup features

### ISSUE THREAD

**Issue by refly2119** [PufferPanel/PufferPanel#842]
Please use the [community forums](https://community.pufferpanel.com) for issues, questions, or help with installation or configuration of PufferPanel.
This issue tracker is only for confirmed bug reports. Please consider opening a thread on the community forums before opening an issue here.
Please delete these 3 lines before submitting your issue.

**Comment by InflamedSebi** [comment]
Since this is a topic already:
Just want to draw 'rdiff-backup' to your attention.
Normal backups are 'hard' to manage if you don't want to waste tons of space using only full backups, so you need incremental/differential backups.
This needs some basic knowledge on how these actually work and you cannot simply prune anything older than X , since you have to keep the full backup, and the corresponding incremental/differential backups for this point in time etc.
While this is tricky to implement in an intuitive way for non IT people, there is a better approach to this: 'Reverse Differential Backups'
This means you have only a single full backup and everything else is differential. The key is the reverse aspect here.
The full backup is always the most recent backup and every older backup is a differential one to the next newer backup.
BN(diff to BN-1) <- ... <- B3(diff to B2) <- B2(diff to B1) <- B1(diff to B0) <- B0(full) <- LiveFiles

Pros of this:
- you can easily prune old backups by specifying either a time span or a number of backups to keep, without worrying to break your backup
- probably one of the most space preserving backups
- you don't need knowledge in how backups work, to use it
- restoring recent data is faster
- the most recent backup can be restored via copy & paste since it is a mirror
Cons:
- restoring very old data can be slower

For this explicit project a cronjob & button executing 'rdiff-backup -b --exclude ${pathToServer}/backup ${pathToServer} ${pathToServer}/backup' will create backups of that server inside a relative backup directory, but omitting the backup directory itself to prevent infinite loops. Then doing 'rdiff-backup --remove-older-than 20B ${pathToServer}/backup' will prune all older backups than 20 differential backups, but specifying a time like 6 weeks will work too 'rdiff-backup --remove-older-than 6W ${pathToServer}/backup'.
Of course setting the cronjob and prune options would need a proper place in the webui.
Restoring is quite simple too:
'rdiff-backup -r 3B ${pathToServer}/backup ${pathToServer}' will restore the server based on the 3rd differential backup (0B would be the very last backup). Or you can specify '-r 3D5h' to get the version 3 days and 5 hours ago. If there was no exact backup at this time, it will use the the next older one to the time specified. Common date formats like '-r 2002-01-25T07:00:00+02:00' will work too. getting a list of all backups with sizes is 'rdiff-backup -l --list-increment-sizes'. Translating that output to a webpage with buttons to restore a target is probably most of the work.

Hope this might help you here :)

**Comment by b-reich** [comment]
A simple Archive (zip/tar.gz) would be great.

**Comment by FriendlyGamer** [comment]
> A simple Archive (zip/tar.gz) would be great.

It's 2020 that's the _worst possible way_ to back up. Rdiff or similar is the way to go typically. Even though disk space prices had somewhat came down, it still not uncommon to pay 3 to 10 cents per GB for small quantities of storage space. That is RAID/Ceph protected or what not.

**Comment by b-reich** [comment]
I understand that. But sometimes full backups are easier to handle instead of incremental ones. You can also take a look at "borg".
And every backup function is better than no backup.

**Comment by FriendlyGamer** [comment]
> I understand that. But sometimes full backups are easier to handle instead of incremental ones.  
Not at all, good luck getting the drive space and network throughput needed and to assure there enough and so on. I rather set Rdiff once and "forget it" knowing that I don't need to install additional storage faster than I can keep up. 
> And every backup function is better than no backup. 
Not at all, proper backing up is just as important as having them.
